src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/WithdrawableTreeBuilder.kt

```diff
@@ -0,0 +1,124 @@
+
+    placementRequestService.getPlacementRequestForInitialApplicationDates(application.id)?.let {
+      children.add(treeForPlacementReq(it, user))
+    }
+
+    placementApplicationService.getAllPlacementApplicationEntitiesForApplicationId(application.id).forEach {
+      children.add(treeForPlacementApp(it, user))
+    }
+
+    bookingService.getAllAdhocForApplication(application).forEach {
+      children.add(treeForBooking(it, user))
+    }
+
+    return WithdrawableTreeNode(
+      entityType = WithdrawableEntityType.Application,
+      entityId = application.id,
+      status = applicationService.getWithdrawableState(application, user),
+      dates = emptyList(),
+      children = children,
+    )
+  }
+
+  private fun treeForPlacementApp(placementApplication: PlacementApplicationEntity, user: UserEntity): WithdrawableTreeNode {
+    val children = placementApplication.placementRequests.map { treeForPlacementReq(it, user) }
+
+    return WithdrawableTreeNode(
+      entityType = WithdrawableEntityType.PlacementApplication,
+      entityId = placementApplication.id,
+      status = placementApplicationService.getWithdrawableState(placementApplication, user),
+      dates = placementApplication.placementDates.map { WithdrawableDatePeriod(it.expectedArrival, it.expectedDeparture()) },
+      children = children,
+    )
+  }
+
+  private fun treeForPlacementReq(placementRequest: PlacementRequestEntity, user: UserEntity): WithdrawableTreeNode {
+    val children = listOfNotNull(
+      placementRequest.booking?.let {
+        treeForBooking(it, user)
+      },
+    )
+
+    return WithdrawableTreeNode(
+      entityType = WithdrawableEntityType.PlacementRequest,
+      entityId = placementRequest.id,
+      status = placementRequestService.getWithdrawableState(placementRequest, user),
+      dates = listOf(WithdrawableDatePeriod(placementRequest.expectedArrival, placementRequest.expectedDeparture())),
+      children = children,
+    )
+  }
+
+  private fun treeForBooking(booking: BookingEntity, user: UserEntity): WithdrawableTreeNode {
+    return WithdrawableTreeNode(
+      entityType = WithdrawableEntityType.Booking,
+      entityId = booking.id,
+      status = bookingService.getWithdrawableState(booking, user),
+      dates = listOf(WithdrawableDatePeriod(booking.arrivalDate, booking.departureDate)),
+      children = emptyList(),
+    )
+  }
+}
+
+data class WithdrawableTreeNode(
+  val entityType: WithdrawableEntityType,
+  val entityId: UUID,
+  val status: WithdrawableState,
+  val dates: List<WithdrawableDatePeriod> = emptyList(),
+  val children: List<WithdrawableTreeNode> = emptyList(),
+) {
+  fun flatten(): List<WithdrawableTreeNode> {
+    return listOf(this) + collectDescendants()
+  }
+
+  private fun collectDescendants(): List<WithdrawableTreeNode> {
+    val result = mutableListOf<WithdrawableTreeNode>()
+    children.forEach {
+      result.add(it)
+      result.addAll(it.collectDescendants())
```

Comment:
To be honest I felt that it would be exercised quite heavily by the WithdrawalTest (integration test) so it felt overkill, although in retrospect I think it would be good as a belt-and-braces approach to testing. I have another PR following up this one, i'll look at adding a unit test once that's merged if that's ok?

Reply:
yes of course. It might be nice and sufficient to have an expectation on the output of `toString()`, especially if the expectation could be presented in a readable way. Maybe it's "as simple as" something like:

```kt
assertThat(tree.toString()).isEqualTo(
        """
          Application(295), withdrawable:true
          ---> PlacementRequest(3b4), withdrawable:true
          ------> Booking(a67), withdrawable:false
          ---> PlacementApplication(580), withdrawable:true
          ------> PlacementRequest(d28), withdrawable:true
          ---------> Booking(9fd), withdrawable:true
          ------> PlacementRequest(e07), withdrawable:true
          ---> PlacementApplication(56f), withdrawable:true
          ---> Booking(c00), withdrawable:true
          ---> Booking(836), withdrawable:false 
        """
)
```

Reply:
oh that's lovely, and it tests the toString()!